File: Happyfamily/Assets/Scripts/Gameplay/TidyUpBoard.cs
using System;
using System.Collections.Generic;
using HappyFamily.Data;
using UnityEngine;

namespace HappyFamily.Gameplay
{
    [Serializable]
    public class StackedTile
    {
        public int Index;
        public string ItemId;
        public string DisplayName;
        public int Layer;        // 0 = bottom, higher = on top
        public int GridX;
        public int GridY;
        public bool IsRemoved;

        public List<int> BlockedBy = new List<int>();  // Indices of tiles blocking this one
    }

    [Serializable]
    public class CollectionSlot
    {
        public string ItemId;
        public string DisplayName;
        public bool IsEmpty => string.IsNullOrEmpty(ItemId);

        public void Clear()
        {
            ItemId = null;
            DisplayName = null;
        }

        public void Set(string itemId, string displayName)
        {
            ItemId = itemId;
            DisplayName = displayName;
        }
    }

    public enum TidyUpMoveState
    {
        None,
        Picked,
        Matched,
        SlotsFull
    }

    public readonly struct TidyUpMoveResult
    {
        public TidyUpMoveResult(TidyUpMoveState state, string itemName, int matchCount = 0)
        {
            State = state;
            ItemName = itemName;
            MatchCount = matchCount;
        }

        public TidyUpMoveState State { get; }
        public string ItemName { get; }
        public int MatchCount { get; }
    }

    public class TidyUpBoard
    {
        public const int SlotCount = 7;
        public const int MatchRequired = 3;

        private TidyUpBoard(List<StackedTile> tiles, CollectionSlot[] slots)
        {
            Tiles = tiles;
            Slots = slots;
        }

        public List<StackedTile> Tiles { get; }
        public CollectionSlot[] Slots { get; }

        public bool IsCompleted => CountRemainingTiles() == 0;
        public bool IsFailed => IsSlotsFull() && !HasPendingMatch();

        public static TidyUpBoard Create(TidyUpLevelDefinition levelDefinition)
        {
            var tiles = new List<StackedTile>();
            var tileIndex = 0;

            // Create tiles based on level definition
            foreach (var itemDef in levelDefinition.Items)
            {
                // Each item appears 3 times (for match-3)
                for (var copy = 0; copy < MatchRequired; copy++)
                {
                    tiles.Add(new StackedTile
                    {
                        Index = tileIndex++,
                        ItemId = itemDef.ItemId,
                        DisplayName = itemDef.DisplayName
                    });
                }
            }

            // Shuffle tiles
            Shuffle(tiles);

            // Assign positions in a stacked grid layout
            AssignStackedPositions(tiles, levelDefinition.GridWidth, levelDefinition.GridHeight, levelDefinition.MaxLayers);

            // Calculate blocking relationships
            CalculateBlocking(tiles);

            // Initialize empty slots
            var slots = new CollectionSlot[SlotCount];
            for (var i = 0; i < SlotCount; i++)
            {
                slots[i] = new CollectionSlot();
            }

            return new TidyUpBoard(tiles, slots);
        }

        public bool CanPickTile(int tileIndex)
        {
            if (tileIndex < 0 || tileIndex >= Tiles.Count)
            {
                return false;
            }

            var tile = Tiles[tileIndex];
            if (tile.IsRemoved)
            {
                return false;
            }

            // Check if blocked by any non-removed tile
            foreach (var blockerIndex in tile.BlockedBy)
            {
                if (!Tiles[blockerIndex].IsRemoved)
                {
                    return false;
                }
            }

            return true;
        }

        public TidyUpMoveResult TryPickTile(int tileIndex)
        {
            if (!CanPickTile(tileIndex))
            {
                return new TidyUpMoveResult(TidyUpMoveState.None, string.Empty);
            }

            var tile = Tiles[tileIndex];

            // Find empty slot
            var slotIndex = FindEmptySlot();
            if (slotIndex < 0)
            {
                return new TidyUpMoveResult(TidyUpMoveState.SlotsFull, tile.DisplayName);
            }

            // Pick the tile
            tile.IsRemoved = true;
            Slots[slotIndex].Set(tile.ItemId, tile.DisplayName);

            // Check for match
            var matchCount = CountMatchingInSlots(tile.ItemId);
            if (matchCount >= MatchRequired)
            {
                RemoveMatchingFromSlots(tile.ItemId);
                return new TidyUpMoveResult(TidyUpMoveState.Matched, tile.DisplayName, matchCount);
            }

            // Check if slots are now full with no pending match
            if (IsSlotsFull() && !HasPendingMatch())
            {
                return new TidyUpMoveResult(TidyUpMoveState.SlotsFull, tile.DisplayName);
            }

            return new TidyUpMoveResult(TidyUpMoveState.Picked, tile.DisplayName);
        }

        public int FindEmptySlot()
        {
            for (var i = 0; i < Slots.Length; i++)
            {
                if (Slots[i].IsEmpty)
                {
                    return i;
                }
            }
            return -1;
        }

        public bool IsSlotsFull()
        {
            for (var i = 0; i < Slots.Length; i++)
            {
                if (Slots[i].IsEmpty)
                {
                    return false;
                }
            }
            return true;
        }

        public bool HasPendingMatch()
        {
            var counts = new Dictionary<string, int>();
            foreach (var slot in Slots)
            {
                if (!slot.IsEmpty)
                {
                    if (!counts.ContainsKey(slot.ItemId))
                    {
                        counts[slot.ItemId] = 0;
                    }
                    counts[slot.ItemId]++;
                    if (counts[slot.ItemId] >= MatchRequired)
                    {
                        return true;
                    }
                }
            }
            return false;
        }

        private int CountMatchingInSlots(string itemId)
        {
            var count = 0;
            foreach (var slot in Slots)
            {
                if (!slot.IsEmpty && slot.ItemId == itemId)
                {
                    count++;
                }
            }
            return count;
        }

        private void RemoveMatchingFromSlots(string itemId)
        {
            var removed = 0;
            for (var i = 0; i < Slots.Length && removed < MatchRequired; i++)
            {
                if (!Slots[i].IsEmpty && Slots[i].ItemId == itemId)
                {
                    Slots[i].Clear();
                    removed++;
                }
            }

            // Compact slots (move items to fill gaps)
            CompactSlots();
        }

        private void CompactSlots()
        {
            var writeIndex = 0;
            for (var readIndex = 0; readIndex < Slots.Length; readIndex++)
            {
                if (!Slots[readIndex].IsEmpty)
                {
                    if (writeIndex != readIndex)
                    {
                        Slots[writeIndex].Set(Slots[readIndex].ItemId, Slots[readIndex].DisplayName);
                        Slots[readIndex].Clear();
                    }
                    writeIndex++;
                }
            }
        }

        private int CountRemainingTiles()
        {
            var count = 0;
            foreach (var tile in Tiles)
            {
                if (!tile.IsRemoved)
                {
                    count++;
                }
            }
            return count;
        }

        private static void AssignStackedPositions(List<StackedTile> tiles, int gridWidth, int gridHeight, int maxLayers)
        {
            var positionIndex = 0;
            var tilesPerLayer = gridWidth * gridHeight;

            foreach (var tile in tiles)
            {
                var layer = positionIndex / tilesPerLayer;
                var posInLayer = positionIndex % tilesPerLayer;

                tile.Layer = Mathf.Min(layer, maxLayers - 1);
                tile.GridX = posInLayer % gridWidth;
                tile.GridY = posInLayer / gridWidth;

                positionIndex++;
            }
        }

        private static void CalculateBlocking(List<StackedTile> tiles)
        {
            // A tile is blocked by tiles on higher layers that overlap its position
            foreach (var tile in tiles)
            {
                tile.BlockedBy.Clear();

                foreach (var other in tiles)
                {
                    if (other.Index == tile.Index)
                    {
                        continue;
                    }

                    // Check if other tile is on a higher layer and overlaps position
                    if (other.Layer > tile.Layer)
                    {
                        // Check for overlap (tiles on higher layers block if they're at same or adjacent position)
                        var dx = Mathf.Abs(other.GridX - tile.GridX);
                        var dy = Mathf.Abs(other.GridY - tile.GridY);

                        // Direct overlap or partial overlap (adjacent with layer difference)
                        if (dx <= 1 && dy <= 1 && (dx + dy) <= 1)
                        {
                            tile.BlockedBy.Add(other.Index);
                        }
                    }
                }
            }
        }

        private static void Shuffle(List<StackedTile> tiles)
        {
            for (var i = tiles.Count - 1; i > 0; i--)
            {
                var j = UnityEngine.Random.Range(0, i + 1);
                (tiles[i], tiles[j]) = (tiles[j], tiles[i]);
            }

            // Re-assign indices after shuffle
            for (var i = 0; i < tiles.Count; i++)
            {
                tiles[i].Index = i;
            }
        }
    }
}

File: minigame/js/gameplay/tidyUpBoard.js
/**
 * 收纳整理游戏逻辑
 * 移植自 Unity TidyUpBoard.cs
 */

import { getLevelItems } from '../data/contentFactory';

// 移动结果状态
export const TidyUpMoveState = {
  None: 'none',
  Picked: 'picked',
  Matched: 'matched',
  SlotsFull: 'slotsFull',
};

// 常量
const SLOT_COUNT = 7;
const MATCH_REQUIRED = 3;

/**
 * 收集槽
 */
class CollectionSlot {
  constructor() {
    this.itemId = -1;
    this.displayName = '';
    this.imageSrc = null;
  }

  get isEmpty() {
    return this.itemId === -1;
  }

  clear() {
    this.itemId = -1;
    this.displayName = '';
    this.imageSrc = null;
  }

  set(itemId, displayName, imageSrc) {
    this.itemId = itemId;
    this.displayName = displayName;
    this.imageSrc = imageSrc;
  }
}

/**
 * 收纳整理游戏板
 */
export default class TidyUpBoard {
  tiles = [];                    // 所有图块
  slots = [];                    // 收集槽（7个）
  gridWidth = 4;
  gridHeight = 4;
  maxLayers = 3;

  constructor() {
    // 初始化收集槽
    this.slots = [];
    for (let i = 0; i < SLOT_COUNT; i++) {
      this.slots.push(new CollectionSlot());
    }
  }

  /**
   * 工厂方法：创建游戏板
   * @param {Object} levelDefinition - 关卡定义
   * @returns {TidyUpBoard}
   */
  static create(levelDefinition) {
    const board = new TidyUpBoard();

    board.gridWidth = levelDefinition.gridWidth || 4;
    board.gridHeight = levelDefinition.gridHeight || 4;
    board.maxLayers = levelDefinition.maxLayers || 3;

    // 获取关卡物品列表
    const items = getLevelItems(levelDefinition);

    // 创建图块（每个物品出现3次，用于三消）
    const allTiles = [];
    items.forEach((item) => {
      for (let i = 0; i < MATCH_REQUIRED; i++) {
        allTiles.push({
          index: -1,  // 稍后分配
          itemId: item.id,
          displayName: item.label,
          imageSrc: item.image,
          layer: 0,
          gridX: 0,
          gridY: 0,
          isRemoved: false,
          blockedBy: [],
        });
      }
    });

    // 打乱顺序
    TidyUpBoard.shuffle(allTiles);

    // 分配索引和位置
    allTiles.forEach((tile, index) => {
      tile.index = index;
    });

    // 分配堆叠位置
    TidyUpBoard.assignStackedPositions(
      allTiles,
      board.gridWidth,
      board.gridHeight,
      board.maxLayers
    );

    // 计算阻挡关系
    TidyUpBoard.calculateBlocking(allTiles);

    board.tiles = allTiles;
    return board;
  }

  /**
   * Fisher-Yates 洗牌算法
   */
  static shuffle(array) {
    for (let i = array.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [array[i], array[j]] = [array[j], array[i]];
    }
    return array;
  }

  /**
   * 分配堆叠位置
   * @param {Array} tiles - 图块列表
   * @param {number} gridWidth - 网格宽度
   * @param {number} gridHeight - 网格高度
   * @param {number} maxLayers - 最大层数
   */
  static assignStackedPositions(tiles, gridWidth, gridHeight, maxLayers) {
    const positionsPerLayer = gridWidth * gridHeight;

    tiles.forEach((tile, index) => {
      tile.layer = Math.min(Math.floor(index / positionsPerLayer), maxLayers - 1);
      const posInLayer = index % positionsPerLayer;
      tile.gridX = posInLayer % gridWidth;
      tile.gridY = Math.floor(posInLayer / gridWidth);
    });
  }

  /**
   * 计算阻挡关系
   * 上层的图块会阻挡下层相邻位置的图块
   */
  static calculateBlocking(tiles) {
    for (const tile of tiles) {
      tile.blockedBy = [];

      // 找出所有在更高层的图块
      const higherTiles = tiles.filter(t =>
        t.layer > tile.layer && !t.isRemoved
      );

      // 检查是否有重叠或相邻的图块
      for (const higher of higherTiles) {
        const dx = Math.abs(higher.gridX - tile.gridX);
        const dy = Math.abs(higher.gridY - tile.gridY);
        const manhattan = dx + dy;

        // 如果在相邻范围内（包括对角线方向的部分重叠）
        if (dx <= 1 && dy <= 1 && manhattan <= 1) {
          tile.blockedBy.push(higher.index);
        }
      }
    }
  }

  /**
   * 检查图块是否可以被拾取
   * @param {number} tileIndex - 图块索引
   * @returns {boolean}
   */
  canPickTile(tileIndex) {
    if (tileIndex < 0 || tileIndex >= this.tiles.length) {
      return false;
    }

    const tile = this.tiles[tileIndex];

    // 已移除的图块不能拾取
    if (tile.isRemoved) {
      return false;
    }

    // 检查是否被其他图块阻挡
    for (const blockerIndex of tile.blockedBy) {
      if (!this.tiles[blockerIndex].isRemoved) {
        return false;
      }
    }

    return true;
  }

  /**
   * 尝试拾取图块
   * @param {number} tileIndex - 图块索引
   * @returns {Object} 移动结果 { state, itemName, matchCount }
   */
  tryPickTile(tileIndex) {
    if (!this.canPickTile(tileIndex)) {
      return { state: TidyUpMoveState.None, itemName: '', matchCount: 0 };
    }

    const tile = this.tiles[tileIndex];

    // 查找空槽位
    const emptySlotIndex = this.findEmptySlot();
    if (emptySlotIndex === -1) {
      return { state: TidyUpMoveState.SlotsFull, itemName: tile.displayName, matchCount: 0 };
    }

    // 移除图块
    tile.isRemoved = true;

    // 将物品放入槽位
    this.slots[emptySlotIndex].set(tile.itemId, tile.displayName, tile.imageSrc);

    // 检查是否达成三消
    const matchCount = this.countMatchingInSlots(tile.itemId);
    if (matchCount >= MATCH_REQUIRED) {
      // 移除匹配的物品
      this.removeMatchingFromSlots(tile.itemId);
      return { state: TidyUpMoveState.Matched, itemName: tile.displayName, matchCount };
    }

    // 检查是否槽位已满且无法消除
    if (this.isSlotsFull() && !this.hasPendingMatch()) {
      return { state: TidyUpMoveState.SlotsFull, itemName: tile.displayName, matchCount: 0 };
    }

    return { state: TidyUpMoveState.Picked, itemName: tile.displayName, matchCount };
  }

  /**
   * 查找空槽位
   * @returns {number} 空槽位索引，无空槽位返回 -1
   */
  findEmptySlot() {
    for (let i = 0; i < this.slots.length; i++) {
      if (this.slots[i].isEmpty) {
        return i;
      }
    }
    return -1;
  }

  /**
   * 检查槽位是否已满
   */
  isSlotsFull() {
    return this.slots.every(slot => !slot.isEmpty);
  }

  /**
   * 检查是否有待消除的匹配
   */
  hasPendingMatch() {
    const counts = {};
    for (const slot of this.slots) {
      if (!slot.isEmpty) {
        counts[slot.itemId] = (counts[slot.itemId] || 0) + 1;
        if (counts[slot.itemId] >= MATCH_REQUIRED) {
          return true;
        }
      }
    }
    return false;
  }

  /**
   * 统计槽位中指定物品的数量
   * @param {number} itemId
   */
  countMatchingInSlots(itemId) {
    return this.slots.filter(slot => slot.itemId === itemId).length;
  }

  /**
   * 从槽位中移除匹配的物品
   * @param {number} itemId
   */
  removeMatchingFromSlots(itemId) {
    let removed = 0;
    for (const slot of this.slots) {
      if (slot.itemId === itemId && removed < MATCH_REQUIRED) {
        slot.clear();
        removed++;
      }
    }
    // 压缩槽位
    this.compactSlots();
  }

  /**
   * 压缩槽位（移除空隙）
   */
  compactSlots() {
    const items = this.slots
      .filter(slot => !slot.isEmpty)
      .map(slot => ({ itemId: slot.itemId, displayName: slot.displayName, imageSrc: slot.imageSrc }));

    // 清空所有槽位
    for (const slot of this.slots) {
      slot.clear();
    }

    // 重新填充
    items.forEach((item, i) => {
      this.slots[i].set(item.itemId, item.displayName, item.imageSrc);
    });
  }

  /**
   * 计算剩余图块数
   */
  countRemainingTiles() {
    return this.tiles.filter(t => !t.isRemoved).length;
  }

  /**
   * 是否完成（所有图块已消除）
   */
  get isCompleted() {
    return this.countRemainingTiles() === 0;
  }

  /**
   * 是否失败（槽位已满且无法消除，且还有图块未消除）
   */
  get isFailed() {
    return this.isSlotsFull() && !this.hasPendingMatch() && !this.isCompleted;
  }

  /**
   * 获取所有可拾取的图块索引
   */
  getPickableTiles() {
    return this.tiles
      .filter((_, index) => this.canPickTile(index))
      .map(tile => tile.index);
  }
}
